internal/graph/storagewrapper.go

```diff
@@ -0,0 +1,346 @@
+package graph
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"golang.org/x/sync/singleflight"
+
+	"github.com/cespare/xxhash/v2"
+	openfgav1 "github.com/openfga/api/proto/openfga/v1"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/trace"
+
+	"github.com/openfga/openfga/internal/build"
+	"github.com/openfga/openfga/internal/keys"
+	"github.com/openfga/openfga/pkg/storage"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+var (
+	// Ensures that Datastore implements the OpenFGADatastore interface.
+	_ storage.OpenFGADatastore = (*CachedDatastore)(nil)
+
+	tuplesCacheTotalCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "tuples_cache_total_count",
+		Help:      "The total number of created cached iterator instances.",
+	})
+
+	tuplesCacheHitCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "tuples_cache_hit_count",
+		Help:      "The total number of cache hits from cached iterator instances.",
+	})
+
+	tuplesCacheDiscardCounter = promauto.NewCounter(prometheus.CounterOpts{
+		Namespace: build.ProjectName,
+		Name:      "tuples_cache_discard_count",
+		Help:      "The total number of discards from cached iterator instances.",
+	})
+
+	tuplesCacheSizeHistogram = promauto.NewHistogram(prometheus.HistogramOpts{
+		Namespace:                       build.ProjectName,
+		Name:                            "tuples_cache_size",
+		Help:                            "The number of tuples cached.",
+		Buckets:                         []float64{1, 10, 100, 1000, 5000, 10000},
+		NativeHistogramBucketFactor:     1.1,
+		NativeHistogramMaxBucketNumber:  100,
+		NativeHistogramMinResetDuration: time.Hour,
+	})
+)
+
+const (
+	QueryCachePrefix = "qc."
+)
+
+// iterFunc is a function closure that returns an iterator
+// from the underlying datastore.
+type iterFunc func(ctx context.Context) (storage.TupleIterator, error)
+
+type CachedDatastore struct {
+	storage.OpenFGADatastore
+
+	cache         storage.InMemoryCache[any]
+	maxResultSize int
+	ttl           time.Duration
+}
+
+// NewCachedDatastore returns a wrapper over a datastore that caches iterators in memory.
+func NewCachedDatastore(
+	inner storage.OpenFGADatastore,
+	cache storage.InMemoryCache[any],
+	maxSize int,
+	ttl time.Duration,
+) *CachedDatastore {
+	return &CachedDatastore{
+		OpenFGADatastore: inner,
+		cache:            cache,
+		maxResultSize:    maxSize,
+		ttl:              ttl,
+	}
+}
+
+// ReadUsersetTuples see [storage.RelationshipTupleReader].ReadUsersetTuples.
+func (c *CachedDatastore) ReadUsersetTuples(
```

Comment:
I probably missed something but why do we not also have implementation for `ReadStartingWithUser`? will it be a follow-up PR?

Reply:
Not for now. I think the cardinality of `ReadStartingWithUser` is not as low to benefit from this iteration. In future work, it might.